### PR TITLE
add all NeoDB features to NiceDB

### PR DESCRIPTION
**Pull request by alphatownsman**
Check List
----------
- [x] create PR
- [x] clean up code
- [x] review db change
- [x] document migration steps
- [x] test run
- [ ] migration
- [ ] remove unused field `is_private`

Migration Steps
---------------
 - Install Redis
 - Follow doc/GUIDE.md, start with the step of `pip install`
 - Migrate visibility field:

```
update books_bookmark set visibility = 2 where is_private = true;
update movies_moviemark set visibility = 2 where is_private = true;
update games_gamemark set visibility = 2 where is_private = true;
update music_albummark set visibility = 2 where is_private = true;
update music_songmark set visibility = 2 where is_private = true;
```
 - Import current entity data into change history:
```
python manage.py populate_history --auto
```
 - Add API key from TMDB and other configurations as needed

### CODE REVIEW

**Comment by doubaniux** [comment]
It is amazing that there is no conflict 👀

**Comment by alphatownsman** [comment]
updated migration steps in description. everything I can think of is added.

You may try with a test site first.

File: boofilsic/settings.py [hunk 1]
```diff
@@ -231,15 +277,30 @@
...
 
 # Spotify credentials
 SPOTIFY_CREDENTIAL = "NzYzNkYTE6MGQ0ODY0NTY2Y2b3n645sdfgAyY2I1ljYjg3Nzc0MjIwODQ0ZWE="
 
 # IMDb API service https://imdb-api.com/
 IMDB_API_KEY = "k23fwewff23"
 
+# The Movie Database (TMDB) API Keys
+TMDB_API3_KEY = "deadbeef"
+TMDB_API4_KEY = "deadbeef.deadbeef.deadbeef"
```
Comment (doubaniux) [inline boofilsic/settings.py:295]: Why are there more than one TMDB keys?

Comment (alphatownsman) [reply boofilsic/settings.py:295]: TMDB has two versions of APIs each using a different key, we are using v3 now, v4 is new but incomplete.

https://developers.themoviedb.org/3/

https://developers.themoviedb.org/4/

File: boofilsic/settings.py [hunk 2]
```diff
@@ -54,7 +62,12 @@
     'movies.apps.MoviesConfig',
     'music.apps.MusicConfig',
     'games.apps.GamesConfig',
+    'sync.apps.SyncConfig',
+    'collection.apps.CollectionConfig',
+    'timeline.apps.TimelineConfig',
     'easy_thumbnails',
+    'user_messages',
+    'django_slack',
```
Comment (doubaniux) [inline boofilsic/settings.py:70]: Maybe a mistake at my side, but I don't see the use of this anywhere else?

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+django-user-messages
+django-slack
+meilisearch
+easy-thumbnails
+lxml
+openpyxl
+psycopg2
+requests
+filetype
+setproctitle
```
Comment (doubaniux) [inline requirements.txt:18]: I don't find the occurrence of this package in the code, was this an attempt to change the process name or get the name of some process?

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+tqdm
+opencc
+dnspython
+typesense
+markdownify
+sentry-sdk
+gitpython
```
Comment (doubaniux) [inline requirements.txt:25]: Same seemingly unused package

File: common/templates/partial/_common_libs.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      release: "NeoDB@{{ version_hash }}",
+      environment: "{{ settings_module }}",
+      tracesSampleRate: 1.0,
+    });
+</script>
+{% endif %}
+{% if jquery %}
+<script src="https://static.neodb.social/cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
+{% else %}
+<script src="https://static.neodb.social/cdnjs.cloudflare.com/ajax/libs/cash/8.1.1/cash.min.js"></script>
```
Comment (doubaniux) [inline common/templates/partial/_common_libs.html:16]: Can cash completely replace jquery in our use case?

File: boofilsic/settings.py (hunk 2 above)
Comment (alphatownsman) [reply boofilsic/settings.py:70]: user_messages is used to send a notification fwhen background tasks completes (e.g. export)

django_slack can be configured in `settings.py` to send notification to an internal instance when 500. I have this in my prod settings: 
```
SLACK_TOKEN = '...'
SLACK_CHANNEL = 'xxx'

LOGGING = {
    'handlers': {
        'slack': {
            'level': 'ERROR',
            'class': 'django_slack.log.SlackExceptionHandler',
        },
    },
```

File: requirements.txt (hunk 3 above)
Comment (alphatownsman) [reply requirements.txt:18]: `django-rq` worker process will have a better title (as displayed by system tools such as ps and top).

File: requirements.txt (hunk 4 above)
Comment (alphatownsman) [reply requirements.txt:25]: typesense and meilisearch are both supported.

File: common/templates/partial/_common_libs.html (hunk 5 above)
Comment (alphatownsman) [reply common/templates/partial/_common_libs.html:16]: not yet user home editor, which I plan to fix later.

File: requirements.txt (hunk 4 above)
Comment (alphatownsman) [reply requirements.txt:25]: and I'm using typesense in prod right now for performance reason.

Comment (alphatownsman) [reply requirements.txt:25]: markdownify is required to import review (html -> md)

Comment (alphatownsman) [reply requirements.txt:25]: sentry-sdk and gitpython is enabled only in my prod settings to send stack track to sentry.io when 500.

File: boofilsic/settings.py (hunk 2 above)
Comment (doubaniux) [reply boofilsic/settings.py:70]: user_messages was accidentally selected, makes sense, thanks

File: requirements.txt (hunk 4 above)
Comment (doubaniux) [reply requirements.txt:25]: I don't why I only wanted to comment on one line but it all turned out to be multiple lines... anyway thank you for the explaination

File: boofilsic/settings.py [hunk 6]
```diff
@@ -231,15 +277,30 @@
...
+# The Movie Database (TMDB) API Keys
+TMDB_API3_KEY = "deadbeef"
+TMDB_API4_KEY = "deadbeef.deadbeef.deadbeef"
+
+# Google Books API Key
+GOOGLE_API_KEY = 'deadbeef-deadbeef-deadbeef'
+
+# IGDB
+IGDB_CLIENT_ID = 'deadbeef'
+IGDB_ACCESS_TOKEN = 'deadbeef'
```
Comment (doubaniux) [inline boofilsic/settings.py:302]: How do you keep this token valid? I think it expires after some time?

**Review by doubaniux — APPROVED** [review]
Thank you for your dedicated contribution, the PR looks good to me :+1:

**Pull request merged**